README.md
# porcupine

## pre-commit setup
This project uses pre-commit, isort, black, and flake8 to help enforce best practices. These libraries are all included in `requirements-dev.txt` and can be installed with pip by running:
```console
pip install -r requirements-dev.txt
```
Once pre-commit is installed, install the hooks specified by the config file into .git:
```console
pre-commit install
```
You can then test pre-commit by running:
```console
pre-commit
```

pre-commit will now run the configured actions as git hooks specified in `.pre-commit-config.yaml`


CHANGELOG.md
## medaprep changelog
-----------

### Version 0.1.1 - 2022-08-11
- Added GitHub actions

### Version 0.1.0 - 2022-08-11
The 0.1.0 release is the first public release available on GitHub and PyPI.
- Added skim.features
- Added visualize.distributions
- Added visualize.query


geolt/tests/test_skim.py
import pytest
import numpy as np
import xarray as xr
import pandas as pd
from pandas.testing import assert_series_equal

from geolt import skim


def create_input_dataset(seed=27) -> xr.Dataset:
    # create test dataset
    np.random.seed(seed)
    elevation = 5 * np.random.randn(2, 2, 3) + 10
    precipitation = 0.03 * np.random.rand(2, 2, 3) + 3
    temperature = 5 * np.random.rand(2, 2, 3) + 30
    precipitation[0, 1, :] *= np.nan
    lon = [[-99.83, -99.32], [-99.79, -99.23]]
    lat = [[42.25, 42.21], [42.63, 42.59]]
    time = pd.date_range("2022-08-01", periods=3)
    reference_time = pd.Timestamp("2022-10-10")

    ds = xr.Dataset(
        data_vars=dict(
            elevation=(["x", "y", "time"], elevation),
            temperature=(["x", "y", "time"], temperature),
            precipitation=(["x", "y", "time"], precipitation),
        ),
        coords=dict(
            lon=(["x", "y"], lon),
            lat=(["x", "y"], lat),
            time=time,
            reference_time=reference_time,
        ),
        attrs=dict(
            description="Weather related data.",
            CRS="EPSG:32615",
            resolution=10.0,
            mu_elevation=10,
            std_elevation=5,
            mu_precipiation=3,
            std_precipitation=0.03,
            mu_temperature=30,
            std_temperature=5,
        ),
    )

    return ds


@pytest.fixture
def dataset_input():
    data = xr.concat(
        [create_input_dataset(seed=s) for s in range(5)],
        pd.Index([s for s in range(5)], name="sensor"),
    )
    return data


def skim_output():
    data = {
        "variables": ["elevation", "temperature", "precipitation"],
        "data_types": ["float64", "float64", "float64"],
        "NaNs": [False, False, True],
        "mean": [10.0, 30.0, 3.0],
        "std": [5.0, 5.0, 3.0],
    }
    df = pd.DataFrame(data)
    df["resolution"] = 10.0
    df["CRS"] = "EPSG:32615"
    return df


def test_skim(dataset_input):
    df_skim = skim(dataset_input)
    target = skim_output()
    assert_series_equal(df_skim["variables"], target["variables"])
    assert_series_equal(df_skim["data_types"], target["data_types"])
    assert_series_equal(df_skim["NaNs"], target["NaNs"])
    assert_series_equal(df_skim["resolution"], target["resolution"])
    assert_series_equal(df_skim["CRS"], target["CRS"])


medaprep/tests/conftest.py
import numpy as np
import pandas as pd
import pytest
import xarray as xr


def create_input_dataset(seed=27) -> xr.Dataset:
    """Construct a test xarray Dataset."""
    np.random.seed(seed)
    elevation = 5 * np.random.randn(2, 2, 3) + 10
    precipitation = 0.03 * np.random.rand(2, 2, 3) + 3
    temperature = 5 * np.random.rand(2, 2, 3) + 30
    precipitation[0, 1, :] *= np.nan
    lon = [[0.0, 100.0], [200.0, 300.0]]
    lat = [[0.0, 100.0], [200.0, 300.0]]
    time = pd.date_range("2022-08-01", periods=3)
    reference_time = pd.Timestamp("2022-10-10")

    ds = xr.Dataset(
        data_vars=dict(
            elevation=(["x", "y", "time"], elevation),
            temperature=(["x", "y", "time"], temperature),
            precipitation=(["x", "y", "time"], precipitation),
        ),
        coords=dict(
            lon=(["x", "y"], lon),
            lat=(["x", "y"], lat),
            time=time,
            reference_time=reference_time,
        ),
        attrs=dict(
            description="Weather related data.",
            spatial_ref="EPSG:32615",
            resolution=10.0,
            mu_elevation=10,
            std_elevation=5,
            mu_precipiation=3,
            std_precipitation=0.03,
            mu_temperature=30,
            std_temperature=5,
        ),
    )

    return ds


@pytest.fixture(name="test_dataset", scope="session")
def fixture_test_dataset():
    """Test fixture to create a test input dataset."""
    data = create_input_dataset()
    return data



medaprep/tests/test_skim.py
# -*- coding: utf-8 -*-
"""medaprep test suite for skim module.

This module implements the test suite for medaprep's skim module.

This module is intended to be executed by running pytest from the project's
root directory.

"""
import numpy as np
import pandas as pd
import pytest
import xarray as xr
from medaprep import skim
from pandas.testing import assert_series_equal


def skim_output():
    """Create the output we expect from running skim to test against."""
    data = {
        "variables": ["elevation", "temperature", "precipitation"],
        "data_types": ["float64", "float64", "float64"],
        "NaNs": [False, False, True],
        "mean": [10.0, 30.0, 3.0],
        "std": [5.0, 5.0, 3.0],
    }
    df = pd.DataFrame(data)
    df["resolution"] = 10
    df["CRS"] = "EPSG:32615"
    return df


def test_skim_features(test_dataset):
    """Test that skimming the constructed input matches the expected output."""
    target = skim_output()
    skim_df = skim.features(test_dataset)
    assert_series_equal(skim_df["variables"], target["variables"])
    assert_series_equal(skim_df["data_types"], target["data_types"])
    assert_series_equal(skim_df["NaNs"], target["NaNs"])


@pytest.fixture(name="test_skim_dataset")
def skim_memory_input():
    """Create a sample input for skim memory."""
    temp = 15 + 8 * np.random.randn(2, 2, 3)
    precip = 10 * np.random.rand(2, 2, 3)
    lon = [[-99.83, -99.32], [-99.79, -99.23]]
    lat = [[42.25, 42.21], [42.63, 42.59]]
    ds = xr.Dataset(
            {
                "temperature": (["x", "y", "time"], temp),
                "precipitation": (["x", "y", "time"], precip),
                },
            coords={
                "lon": (["x", "y"], lon),
                "lat": (["x", "y"], lat),
                "time": pd.date_range("2014-09-06", periods=3),
                "reference_time": pd.Timestamp("2014-09-05"),
                },
            )
    return ds


def skim_memory_output():
    """Create the expected output from skim memory."""
    data = {
            "Index": 577,
            "temperature": 96,
            "precipitation": 96,
            "lon": 96,
            "lat": 96,
            "reference_time": 96
            }

    s = pd.Series(data)
    return s


def test_skim_memory(test_skim_dataset):
    """Test skimming xarray dataset and returning memory in bytes."""
    target = skim_memory_output()
    skim_memory_df = skim.memory(test_skim_dataset)
    assert_series_equal(target, skim_memory_df)


@pytest.fixture(name="test_skim_dask_dataset")
def skim_memory_dask_input():
    """Create sample dask input for skim memory."""
    temp = 15 + 8 * np.random.randn(2, 2, 3)
    precip = 10 * np.random.rand(2, 2, 3)
    lon = [[-99.83, -99.32], [-99.79, -99.23]]
    lat = [[42.25, 42.21], [42.63, 42.59]]
    ds = xr.Dataset(
            {
                "temperature": (["x", "y", "time"], temp),
                "precipitation": (["x", "y", "time"], precip),
                },
            coords={
                "lon": (["x", "y"], lon),
                "lat": (["x", "y"], lat),
                "time": pd.date_range("2014-09-06", periods=3),
                "reference_time": pd.Timestamp("2014-09-05"),
                },
            )
    ds = ds.chunk(chunks={})
    return ds


def skim_memory_dask_output():
    """Create the expected output from skim memory for dask dataset."""
    data = {
            "Index": 128,
            "lat": 96,
            "lon": 96,
            "precipitation": 96,
            "reference_time": 96,
            "temperature": 96,
            "time": 96,
            "x": 96,
            "y": 96,
            }
    s = pd.Series(data)
    return s


def test_skim_dask_memory(test_skim_dask_dataset):
    """Test skimming xarray dask dataset and returning memory in bytes."""
    target = skim_memory_dask_output()
    skim_memory_df = skim.memory(test_skim_dask_dataset)
    assert_series_equal(target, skim_memory_df)



medaprep/visualize.py
# -*- coding: utf-8 -*-
"""Visualize module for use with data processed by medaprep.

This module implements visualization functionality that enables displaying
the results of data processing outputs from medaprep.

This module is part of the core medaprep library and is intended to be called
by user code.

"""
import colorcet as cc
import folium
import numpy as np
import pandas as pd
import xarray as xr
from bokeh.models import ColumnDataSource
from bokeh.plotting import figure
from scipy.stats import gaussian_kde
from shapely.geometry import box


def _convert_bounds(bbox: tuple) -> tuple:
    x1, y1, x2, y2 = bbox
    return ((y1, x1), (y2, x2))


def _compute_bounds(bboxs: list[tuple]) -> tuple:
    xs = []
    ys = []
    for b in bboxs:
        xs.append((b[0], b[1]))
        ys.append((b[1], b[3]))
    x1 = np.min(xs)
    x2 = np.max(xs)
    y1 = np.min(ys)
    y2 = np.max(ys)
    return (x1, y1, x2, y2)


def _compute_center(bbox):
    x1, y1, x2, y2 = bbox
    mu_x = (x1 + x2) / 2
    mu_y = (y1 + y2) / 2
    return (mu_y, mu_x)


def query(
    bbox: [tuple | list[tuple]],
    name: [str | list[str]],
    folium_map: folium.Map,
    color: [str | list[str]],
) -> folium.Map:
    """
    query takes in a list of bounding boxes (bbox), a list of names
    corresponding to the bounding boxes (name), and a folium map (m).
    It adds the bounding boxes to the map (m) with the corresponding
    names, and colors the boxes based on the list of colors (color). It then
    sets the bounds of the map based on the largest provided bounding box, and
    returns the map.

    Args:
        bbox (tuple(s)): containing (x1, y1, x2, y2) latitude and longitude
            coordinates of bounding boxes.
        name (str): containing a name for each bbox.
        m (folium.Map): map to plot boxes on.
        color (str): color for each bounding box.

    Returns:
        folium.Map containing bounding boxes

    """
    if not isinstance(bbox, list):
        bbox = [bbox]
    if not isinstance(name, list):
        name = [name]
    if not isinstance(color, list):
        color = [color]
    assert len(bbox) == len(name) == len(color)

    for (b, n, c) in zip(bbox, name, color):
        folium.GeoJson(
            box(*b),
            style_function=lambda x, color=c: dict(
                fill=False, weight=5, opacity=0.5, color=color
            ),
            name=n + " Center",
            tooltip=n,
        ).add_to(folium_map)

        folium.Marker(
            _compute_center(b),
            popup=n + " Center",
            icon=folium.Icon(color=c, icon="star"),
        ).add_to(folium_map)

    folium_map.fit_bounds(bounds=_convert_bounds(bbox[0]))
    return folium_map


def distributions(
    indata: xr.Dataset, skim_table: pd.DataFrame, sample_size: int
) -> list:
    """
    This function returns a list of bokeh figures containing
    estimated distributions of the variables within the
    dataset.

    Args:
        indata (xarray.Dataset): input data containing variables which
            distributions will be estimated from.
        skim_table (pandas.DataFrame): dataframe containing basic info about
            the dataset.

    Returns:
        Bokeh figures containing estimated distributions of each variable.

    """
    est_pdfs = []
    epsilon = 0.00001
    palette = [cc.rainbow[i * 15] for i in range(17)]
    for i, row in skim_table.iterrows():
        variable = row["variables"]
        min_val = row["minimums"]
        max_val = row["maximums"]
        data_np = indata[variable].to_dataframe()[variable].values
        if row["NaNs"]:
            data_np = data_np[~np.isnan(data_np)]
        pdf = gaussian_kde(data_np)
        x = np.linspace(int(min_val - 5.0), int(max_val + 20.0), sample_size)
        y = pdf(x)
        y /= np.sum(y) + epsilon
        p = figure(
            title=variable,
            height=300,
            width=900,
            x_range=(int(min_val - 5.0), int(max_val + 10.0)),
            toolbar_location="above",
            height_policy="auto",
        )
        source = ColumnDataSource(data=dict(x=x, y=y))
        p.patch(
            x="x",
            y="y",
            color=palette[i],
            alpha=0.6,
            line_color="black",
            source=source,
        )
        p.title.text_font_size = "25px"
        est_pdfs.append(p)
    return est_pdfs


medaprep/skim.py
# -*- coding: utf-8 -*-
"""Skim module.

This module implements utility functions that provide high level details of
raster data.

This module is part of the core medaprep library and is intended to be called
by user code.

"""
import pandas as pd
import xarray as xr


def _dask_skim_memory(indata: xr.Dataset) -> pd.Series:
    """Skim memory of dask-backed Xarray Dataset."""
    ddf = indata.to_dask_dataframe()
    s = ddf.memory_usage().compute()
    return s

def _skim_memory(indata: xr.Dataset) -> pd.Series:
    """Skim memory of Xarray Dataset."""
    df = indata.to_dataframe()
    s = df.memory_usage()
    return s


def memory(indata: xr.Dataset) -> pd.Series:
    """
    This function uses utilities from pandas and dask (for dask-backed datasets)
    to check the memory size of the input dataset.

    Args:
        indata (xarray.Dataset): dataset to be skimmed.

    Returns:
        (pandas.Series): series containing number of bytes for each column.

    Example:

        .. code-block:: python

            >>> print(data)
                <xarray.Dataset>
                Dimensions:      (time: 1, y: 1142, x: 1137)
                Coordinates:
                    * y            (y) float64 3.716e+06 3.715e+06 ... 3.351e+06 3.351e+06
                    * x            (x) float64 -1.102e+07 -1.102e+07 ... -1.066e+07 -1.066e+07
                    spatial_ref  int32 3857
                    * time         (time) datetime64[ns] 2022-07-03T17:25:22
                Data variables:
                    visual       (time, y, x) uint8 dask.array<chunksize=(1, 1142, 1137), meta=np.ndarray>
                    B01          (time, y, x) uint16 dask.array<chunksize=(1, 1142, 1137), meta=np.ndarray>
            >>> print(skim.memory(data))
            Index          6576899
            visual         1298454
            B01            2596908
            spatial_ref    5193816
            dtype: int64

    """
    # check if input is xarray dataset
    assert isinstance(indata, xr.Dataset)

    # check if input is a dask xarray dataset
    if indata.chunks:
        return _dask_skim_memory(indata)
    else:
        return _skim_memory(indata)

def features(indata: xr.Dataset) -> pd.DataFrame:
    """
    This function returns a dataframe with information about the variables,
    data types, null values, means, standard deviations, maximums, and
    minimums for a given dataset.

    Args:
        indata (xarray.Dataset): datset to be skimmed.

    Returns:
        (pandas.DataFrame): table containing basic information about the
        dataset.

    Example:

        .. code-block:: python

            >>> import numpy as np
            >>> import pandas as pd
            >>> import xarray as xr
            >>> from medaprep import skim

            >>> temp = 15 + 8 * np.random.randn(2, 2, 3)
            >>> precip = 10 * np.random.rand(2, 2, 3)
            >>> lon = [[-99.83, -99.32], [-99.79, -99.23]]
            >>> lat = [[42.25, 42.21], [42.63, 42.59]]
            >>> ds = xr.Dataset(
                     {
                         "temperature": (["x", "y", "time"], temp),
                         "precipitation": (["x", "y", "time"], precip),
                         },
                     coords={
                         "lon": (["x", "y"], lon),
                         "lat": (["x", "y"], lat),
                         "time": pd.date_range("2014-09-06", periods=3),
                         "reference_time": pd.Timestamp("2014-09-05"),
                         },
                     )

            >>> df = skim.features(ds)
            >>> df
                variables       data_types  NaNs    mean    std     maximums    minimums
            0   temperature     float64     False   14.3177 9.08339 30.3361     -7.76803
            1   precipitation   float64     False   4.62568 3.03081 9.89768     0.147005

    """
    variables = list(indata.data_vars)
    d = dict(indata.dtypes)
    types = [d[v] for v in list(indata.data_vars)]
    null_counts = indata.isnull().sum().to_array().data > 0
    mus = indata.mean().to_array().data
    stds = indata.std().to_array().data
    maxs = indata.max().to_array().data
    mins = indata.min().to_array().data

    data_dict = {
        "variables": variables,
        "data_types": types,
        "NaNs": null_counts,
        "mean": mus,
        "std": stds,
        "maximums": maxs,
        "minimums": mins,
    }
    df = pd.DataFrame(data_dict)
    return df
